parsons/targetsmart/targetsmart_smartmatch.py

```diff
@@ -0,0 +1,311 @@
+"""Implements client routine to allow execution of TargetSmart SmartMatch
+workflows.
+
+TargetSmart SmartMatch API doc:
+https://docs.targetsmart.com/developers/tsapis/v2/service/smartmatch.html
+"""
+
+import gzip
+import logging
+import shutil
+import tempfile
+import time
+import uuid
+
+import petl
+import requests
+
+logger = logging.getLogger(__name__)
+
+VALID_FIELDS = [
+    "voterbase_id",
+    "smartvan_id",
+    "voter_id",
+    "exact_track",
+    "full_name",
+    "first_name_combined",
+    "first_name",
+    "middle_name",
+    "last_name",
+    "name_suffix",
+    "address1",
+    "address2",
+    "city",
+    "state",
+    "zip",
+    "age",
+    "gender",
+    "dob",
+    "phone",
+    "email",
+    "latitude",
+    "longitude",
+]
+
+INTERNAL_JOIN_ID = "matchback_id"
+INTERNAL_JOIN_ID_CONFLICT = "__matchback_id"
+
+
+class SmartMatchError(Exception):
+    """Raised when SmartMatch workflow processing fails."""
+
+
+def _smartmatch_upload(url, fname):
+    logger.info(f"Uploading {fname} to {url} to begin SmartMatch workflow execution.")
+    with open(fname, "rb") as reader:
+        response_2 = requests.put(url, data=reader, headers={"Content-Type": ""})
+
+    response_2.raise_for_status()
+
+
+def _smartmatch_download(url, writer):
+    with requests.get(url, stream=True) as response:
+        response.raise_for_status()
+        for chunk in response.iter_content(chunk_size=8192):
+            writer.write(chunk)
+
+
+def _add_join_id(input_table):
+    """`matchback_id` is added to the raw input table so the results can later be
+    joined back. Integer sequence values are used. If the column already exists
+    in the raw input, it is renamed to `__matchback_id` and restored after
+    result join.
+    """
+    if INTERNAL_JOIN_ID in input_table.fieldnames():
+        input_table = input_table.rename(INTERNAL_JOIN_ID, INTERNAL_JOIN_ID_CONFLICT)
```

Comment:
These are petl methods, but you are accessing a Parsons table. To access the petl object is a accessed via the `table` attribute of a Parsons Table. 

Ex:

`input_table = input_table.table.rename(INTERNAL_JOIN_ID, INTERNAL_JOIN_ID_CONFLICT)`